replica/table.cc

```diff
@@ -1853,33 +1853,34 @@ void table::set_hit_rate(gms::inet_address addr, cache_temperature rate) {
     e.last_updated = lowres_clock::now();
 }
 
-table::cache_hit_rate table::get_hit_rate(gms::gossiper& gossiper, gms::inet_address addr) {
-    auto it = _cluster_cache_hit_rates.find(addr);
+table::cache_hit_rate table::get_hit_rate(gms::gossiper* gossiper, gms::inet_address addr) {
     if (utils::fb_utilities::get_broadcast_address() == addr) {
         return cache_hit_rate { _global_cache_hit_rate, lowres_clock::now()};
     }
-    if (it == _cluster_cache_hit_rates.end()) {
-        // no data yet, get it from the gossiper
-        auto* eps = gossiper.get_endpoint_state_for_endpoint_ptr(addr);
-        if (eps) {
-            auto* state = eps->get_application_state_ptr(gms::application_state::CACHE_HITRATES);
-            float f = -1.0f; // missing state means old node
-            if (state) {
-                sstring me = format("{}.{}", _schema->ks_name(), _schema->cf_name());
-                auto i = state->value.find(me);
-                if (i != sstring::npos) {
-                    f = strtof(&state->value[i + me.size() + 1], nullptr);
-                } else {
-                    f = 0.0f; // empty state means that node has rebooted
-                }
-                set_hit_rate(addr, cache_temperature(f));
-                return cache_hit_rate{cache_temperature(f), lowres_clock::now()};
+
+    auto it = _cluster_cache_hit_rates.find(addr);
+    if (it != _cluster_cache_hit_rates.end()) {
+        return it->second;
+    }
+
+    // no data yet, get it from the gossiper
+    auto* eps = gossiper ? gossiper->get_endpoint_state_for_endpoint_ptr(addr) : nullptr;
```

Comment:
Why is it safe? It looks like a bad interface. If a query that calls this is distributed the gossiper has to be non null, and if it is not there is no point calling the function.

Reply:
The function is unfortunately used for local node too (first `if`).
I guess I could do some other refactors so we check gossiper presence at call sites.
(TBH those gossiper* commits belong more to a separate PR where I actually move remote init later)

Reply:
The function obviously needs to handle local node as well, but the caller does not have to call this function for local keyspaces. It does now because there is no harm, but if you have to make the interface unsafe we may want to change it.

Reply:
BTW. if it's unsafe to call the function without gossiper, maybe it should also be unsafe to call it if we don't have `endpoint_state` for the node?
Couldn't it happen that we call this function before we gossiped with the node `addr` - so we don't have CACHE_HITRATE for this node yet?
Maybe it's safe. I don't get how that works though, because the function will cache the result and never access gossiper again? I don't get - CACHE_HITRATE is constantly being updated, no? Maybe there's something which clears this `_cluster_cache_hit_rates` local cache.
And if that's safe - then maybe it's also safe if gossiper is missing for a while but then comes back. So the function doesn't really need for gossiper to be present all the time, only eventually, it's not critical that we always calculate the hit rates correctly.

Reply:
It should not happen, but if it happens because of some bug we prefer to misdirect the traffic than fail a read. How a gossiper can be missing for a while?

Reply:
I don't think it could be missing, just wondering how "bad" the interface is, i.e. if the implementation should really consider it a disaster if gossiper is gone.

Reply:
Maybe it could become missing for a while when we're shutting down, or starting up (and there's a race so someone already calls `get_hit_rate` before gossiper becomes available)

Reply:
Interface is bad because the correctness of the return value depends on one of the parameters to be not null. Why would you call with null then? Either do not call it because a query is local or fail if the query is distributed by the gossiper is not available.

Reply:
> correctness of the return value

that's what I'm debating - if the returned hit rate only requires to be eventually consistent (because e.g. endpoint state for a node may be missing for a while) then it's not incorrect to return a default value for some time.

Reply:
When you are debating something please do not ignore previous points made here: https://github.com/scylladb/scylla/pull/11088#discussion_r925725998

Second after you changes it may be that the only possibility is that the gossiper will be null only for a while (which is still a bug for distributed query, why do you even imply it may be OK?), but tomorrow someone will call this function for some other use case and will pass constant null there because he will not have access to the gossiper. The interface is bad because it allows it. For this function the gossiper is not optional. If you want to call it without the gossiper you are doings something wrong.

Reply:
> The function is unfortunately used for local node too (first if).

But isn't this `if` an early return? It looks like the rest of the code only runs for the remote communication case, so you could just call `on_internal_error` if there's no gossiper available.

Reply:
@kbr- , in this patch gossiper is always non-nullptr, where do you call it with nullptr?

Reply:
Removed these controversial `gossiper*` commits.